lib/SymbolGraphGen/SymbolGraphASTWalker.cpp

```diff
@@ -173,30 +190,35 @@ bool SymbolGraphASTWalker::walkToDeclPre(Decl *D, CharSourceRange Range) {
         HandleProtocolOrComposition(InheritedTy);
       }
 
-      while (!UnexpandedCompositions.empty()) {
-        const auto *Comp = UnexpandedCompositions.pop_back_val();
-        for (const auto &Member : Comp->getMembers()) {
-          HandleProtocolOrComposition(Member);
+      while (!UnexpandedCompositions.empty() || !UnexpandedProtocols.empty()) {
```

Comment:
Maybe I misjudged the situation, but I don't think this can be split up into two loops.

What I'm trying to achieve here is to record `conformsTo` relationships between the extension block symbol and any conformances added via this extension. However, we do not only want to include those protocols that are explicitly stated on the extension block, _but also implicitly inherited ones_. To do so, we have to expand two different constructions:

Protocol declarations (`UnexpandedProtocols`), which state one or more inheritances, e.g.:
```swift
public protocol X: A, B {
    // ...
}
```
Protocol compositions (`UnexpandedCompositions`), e.g.:
```swift
public typealias Y = C & D
```

I assume that these two constructions can be nested indefinitely, e.g. like this:

```swift
public protocol A: B, C { }

public typealias B = D & E

public protocol D: F, G { }

public typealias F = H & I

// ...
```

If I were to break up the loop into two separate ones, I couldn't handle this situation, could I?

Reply:
Resolved in bb78034edf9d004b260cf890998699faab7dffba.

Reply:
I still think this is unintuitive control flow, it took me a while to figure out how this was working. Would it not make more sense to keep track only of unexpanded protocols and handle all compositions eagerly? Essentially what I amsuggesting is to make this just an iteration over `UnexpandedProtocols` and turn `HandleProtocolOrComposition` into:
```
// Add the type to the right list for expansion.
      auto HandleProtocolOrComposition = [&](Type InitialTy) {
        SmallVector<const ProtocolCompositionType *, 4> UnexpandedLocalCompositions = { InitialTy };
        while (const auto *Ty = UnexpandedLocalCompositions.pop_back_val()) {
            if (const auto *Proto = dyn_cast_or_null<ProtocolDecl>(Ty->getAnyNominal())) {
                UnexpandedProtocols.push_back(Proto);
            } else if (const auto *Comp = Ty->getAs<ProtocolCompositionType>()) {
                 UnexpandedLocalCompositions.append(Comp->getMembers());
            } else {
               abort();
            }
      };
```
With this you can get rid of `UnexpandedCompositions` at the outer scope and make this loop pretend like it only ever deals with `UnexpandedProtocols` which seems more intuitive.

Reply:
Resolved in 37f4c983c382991edd2c776e98250a9ce3a96d0e. It turned out a little more verbose, though, as I had to add some type assertions in the beginning to make the code compile.